File: Queue/typescript/queue_doubly_linked_list.ts
import * as readline from 'readline'
const rl = readline.createInterface({
    input:process.stdin
})
let input_data:string = ""
rl.on("line",(line)=>{
    input_data+=line
    if(line.includes("-1")){
        rl.close()
    }
})

class Node<T>{
    prev:Node<T>|null
    val:T
    next:Node<T>|null
    constructor(val:T){
        this.prev = null
        this.val = val
        this.next = null
    }
}

class Queue<T>{
    sz:number
    head:Node<T>|null
    tail:Node<T>|null
    constructor(){
        this.sz = 0
        this.head = null
        this.tail = null
    }
    front():T|undefined{
        return this.head?.val
    }

    pop(){
        if (this.head===null){
            return
        }
        this.head = this.head.next
        if(this.head!=null){
            this.head.prev = null
        }
        this.sz--
    }

    push(val:T){
        const newNode = new Node(val)
        if(this.head===null){
            this.head = newNode
            this.tail = newNode
            this.sz++
            return
        }
        if (this.tail){
            this.tail.next = newNode
            newNode.prev = this.tail
            this.tail = newNode
            this.sz++
        }
    }
    size(){
        return this.sz
    }
    isEmpty(){
        return this.size()===0
    }
}

const queue = new Queue<number>()
rl.on("close",()=>{
    const a = input_data.split(" ").map(Number)
    let i = 0;
    while(true){
        if(a[i]===-1){
            break
        }else{
            queue.push(a[i])
        }
        i++
    }
    while(!queue.isEmpty()){
        process.stdout.write(queue.front() +  " ")
        queue.pop()
    }
    console.log()
})

File: Queue/python/queue_doubly_linked_list.py
import sys
class Node:
    def __init__(self,val):
        self.prev = None
        self.val  = val
        self.next = None

class Queue:
    def __init__(self):
        self.sz = -1
        self.head = None
        self.tail = None
    
    def front(self):
        return self.head.val if self.head != None else None

    def pop(self):
        if self.head == None:
            return
        delete_node = self.head
        self.head = self.head.next
        if self.head != None:
            self.head.prev = None
        del delete_node
        self.sz-=1
    
    def push(self,val):
        newNode = Node(val)
        if self.head == None:
            self.head = newNode
            self.tail = newNode
            self.sz+=1
            return
        self.tail.next = newNode
        newNode.prev = self.tail
        self.tail = newNode
        self.sz+=1

    def size(self):
        return self.sz
    
    def isEmpty(self):
        return self.sz==-1

input_data = sys.stdin.read().strip().split(" ")
a = list(map(int,input_data))
queue = Queue()
i = 0
while True:
    if a[i]==-1:
        break
    else:
        queue.push(a[i])
    i+=1
while not queue.isEmpty():
    print(queue.front(),end=" ")
    queue.pop()





